aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteModelFieldTypeConverter.java

```diff
@@ -0,0 +1,242 @@
+        } catch (IOException exception) {
+            LOGGER.warn("", exception);
+            return null;
+        }
+    }
+
+    private Object convertCustomTypeToTarget(Cursor cursor, ModelField field, int columnIndex) throws IOException {
+        final String stringValue = cursor.getString(columnIndex);
+        return gson.getAdapter(field.getType()).fromJson(stringValue);
+    }
+
+    private Object convertDateToTarget(Cursor cursor, ModelField field, int columnIndex) throws ParseException {
+        // TODO wire up the new Date/Time handling here
+        final String dateInStringFormat = cursor.getString(columnIndex);
+        if (dateInStringFormat != null) {
+            return SimpleDateFormat
+                    .getDateInstance()
+                    .parse(dateInStringFormat);
+        }
+        return null;
+    }
+
+    @Override
+    public Object convertValueFromTarget(Model model, ModelField field) throws DataStoreException {
+        final String fieldName = field.getName();
+
+        final Object fieldValue = getModelValue(model, fieldName);
+        if (fieldValue == null) {
+            return null;
+        }
+        final JavaFieldType javaFieldType = TypeConverter.getJavaFieldType(field);
+
+        switch (javaFieldType) {
+            case INTEGER:
+            case LONG:
+            case FLOAT:
+            case STRING:
+                // these types require no special treatment
+                return fieldValue;
+            case BOOLEAN:
+                boolean booleanValue = (boolean) fieldValue;
+                return booleanValue ? 1L : 0L;
+            case MODEL:
+                return ((Model) fieldValue).getId();
+            case ENUM:
+            case CUSTOM_TYPE:
+                return gson.toJson(fieldValue);
+            case DATE:
+                // TODO integrate with new Date/Time handling
+                final Date dateValue = (Date) fieldValue;
+                return SimpleDateFormat
+                        .getDateInstance()
+                        .format(dateValue);
+            case TIME:
+                return ((Time) fieldValue).getTime();
+            default:
+                LOGGER.warn("Field of type " + javaFieldType + " is not supported. Fallback to null.");
+                return null;
+        }
+    }
+
+    private Object getModelValue(Model model, String fieldName) throws DataStoreException {
+        Class<? extends Model> modelClass = model.getClass();
+        final String getterName = "get" + Character.toUpperCase(fieldName.charAt(0)) + fieldName.substring(1);
+        try {
+            final Method fieldGetter = modelClass.getMethod(getterName);
+            return fieldGetter.invoke(model);
+        } catch (Exception exception) {
+            LOGGER.debug("Could not find " + getterName + "() on " + modelClass.getName()
+                    + ". Fallback to direct field access.");
+            // fallback to direct field access
```

Comment:
If we assume that models only come from the code-generator, it might be over-kill to fall back on field inspection, here.

In practice, there's a _whole_ set of expectations our code base has about `M extends Model` type classes. The interface says they need a method called `getId()`, but in reality, we have all of the (currently required) annotations, etc. Should we just assume they have getter methods? Or if we do the field access thing, should we respect the private access of the field? (I guess, tools like GSON don't, they just read the private field ...)

As a note on construction - can we avoid nesting the try/catch in a try/catch here?

Maybe the flow could be like:
```
try {
   return favorableApproach();
} catch (DidntPanOutException favorableDidntWork) {
    // ...
}
try {
    return fallbackApprach();
} catch (FlalbackException alsoDidntWork) {
    // ...
}
```
?

Reply:
I'm not sure we should break in case customers decided (for any reason) to either change the generated model or even to write their own (e.g. with public fields instead, using Lombok or Kotlin data classes). I think we could be resilient here **as long as it's a valid pojo**.

That said, you raise an important question: is a pojo with private fields should have the access level respected as such? I don't know, but the argument that libraries like Gson, Jackson, Spring, etc bypass the access level seems like a strong one to do the same.

Note that we have a few places with direct field access, this is just an initial attempt to standardize it. For the time being I'll refactor it into a more general-purpose helper and we can discuss the rules for field access and also remove field access duplication.

Reply:
Okay, cool. Yes I do agree with you, we should move in the direction of supporting arbitrary `@Model`s, through-out. My argument was basically "But we don't _right now_"; however, that doesn't mean we should keep marching in that direction!